### ISSUE TITLE
SES-140: Disable defer node expansion / SAML HTTP session memory usage

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-security-saml#39]
[Mike Noordermeer](https://jira.spring.io/secure/ViewProfile.jspa?name=miken) (Migrated from [SES-140](https://jira.spring.io/browse/SES-140)) said:

The spring-security-saml samples show the use of a  `StaticBasicParserPool`. The `DocumentBuilder` initialised by the pool is by default using deferred node expansion. Deferred node expansion uses a lot of memory, especially for small documents (like SAML assertions).

Since the `SAMLCredential` is stored in the http session object, this leads to a quite high memory usage (50-100 KB per active session, rooted in a `DeferredDocumentImpl`, referenced by the `SAMLObject.dom` property, which holds the original SAML response). This kills our appserver with only a couple of thousand sessions, using a 1,5GB heap.

Solution seems to be to disable deferred node expansion, like this:

``` xml
<bean id="parserPool" class="org.opensaml.xml.parse.StaticBasicParserPool" init-method="initialize">
     <property name="builderFeatures">
         <map>
             <entry key="http://apache.org/xml/features/dom/defer-node-expansion" value="false"/>
         </map>
     </property>
</bean>
```

The same solutions is recommended by Xerces (http://xerces.apache.org/xerces2-j/faq-performance.html) and seems to be used by OpenSSO (https://java.net/jira/browse/OPENSSO-1322).

**Comment by kostabanderas** [comment]
Even though this an old issue, I have recently dealt with this in our implementation and came across a potential security issue. As described here: https://shibboleth.net/community/advisories/secadv_20131213.txt
Any custom setup of the parserPool is open to vulnerabilities unless the same builderFeatures are provided as in the default constructor of the, let's say, StaticBasicParserPool. Basically, if we do provide any builder feature, default features (which are provided to prevent these vulnerabilities in the first place) are overridden, thus exposing parserPool to those vulnerabilities.
Spcifically, these two features should then also be provided, as can be seen in the default implementation:
features.put("http://javax.xml.XMLConstants/feature/secure-processing", true);
features.put("http://apache.org/xml/features/disallow-doctype-decl", true);

**Comment by fhanik** [comment]
I see what you are saying. When a user decides to override parser features you want the library to still enforce certain attributes in order to avoid accidental disabling of the two default features that are set to true.

I'm going to mark this as not fix because the 1.0.x release train is not being active developed and this is a user configuration issue. The correct work around would be:
```xml
<bean id="parserPool" class="org.opensaml.xml.parse.StaticBasicParserPool" init-method="initialize">
     <property name="builderFeatures">
         <map>
             <entry key="http://apache.org/xml/features/dom/defer-node-expansion" value="false"/>
             <entry key="http://javax.xml.XMLConstants/feature/secure-processing" value="true"/>
             <entry key="http://apache.org/xml/features/disallow-doctype-decl" value="true"/>
         </map>
     </property>
</bean>
```
In the future release, this will be controlled here
https://github.com/spring-projects/spring-security-saml/blob/c49fe77/core/src/main/java/org/springframework/security/saml/spi/opensaml/OpenSamlImplementation.java#L288-L318

If a user decides to override the `bootstrap` method, without calling super, that is done with intention.
